README.md

```diff
@@ -1,4 +1,371 @@
+
+# Assumption
+
+This document is mainly for Azure CLI developers and people who are interested in generating Azure CLI by themselves.  
+# How Does the CLI Code Generator Work.
+The Azure Code Generator is basically a [Autorest](https://github.com/Azure/autorest) extension which takes swagger in [azure-rest-api-specs](https://github.com/Azure/azure-rest-api-specs) or [azure-rest-api-specs-pr](https://github.com/Azure/azure-rest-api-specs-pr) repos as input and output functional Azure CLI code. It uses Autorest V3 to handle the configuration interpretation, pipeline resolving, pipeline scheduling and uses [Autorest.Modelerfour](https://github.com/Azure/autorest.modelerfour) as a basic code model. 
+
+Besides the Autorest.Modelerfour, Autorest.Az has two more Autorest extensions dependencies, the [Autorest.Clicommon](https://github.com/Azure/autorest.clicommon) and [Autorest.Python](https://github.com/Azure/autorest.python), the Autorest.Clicommon mainly handles the user defined cli directives such as split operations, handle polymorphism, rename, hide etc. and properly mark it in the code model, the Autorest.Python is integrated by Autorest.Az as the Azure CLI modules don't call the rest api directly, instead, it either call the vendored SDKs in the case of Azure CLI Extensions or call the public released SDK in the case of Azure CLI main repo modules.
+
+Both the Autorest.Clicommon and Autorest.Python take Autorest.Modelerfour as input, Normally autorest.python will not flatten the code model because the Python Language naturally support complex object, they don't need to flatten the parameter but for command line tools like Azure CLI, the parameter layer must not deeper than 2, otherwise it's hard to express the complex object without using a bunch of delimiters. Therefore, autorest.clicommon will flatten the code model. Autorest az will merge the two code model from autorest.python and autorest.clicommon.    
+
+After autorest.az has done some CLI specific logic, it will render the code model into code template finally output the generated code.   
+
+
+# How to use the CLI Code Generator
+
+Users can use autorest.az to generate the Azure CLI code both in rest api specs PR pipeline and in their local environment. Guidance for generate the Azure CLI code in rest api specs PR pipeline can be found in [rest api specs documentation](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/code-gen/configure-cli.md). Here we only focus on the guidance to generate the CLI code in local.  
+
+More detailed generation guidance can be found in the [how to generate doc](https://github.com/Azure/autorest.az/blob/master/doc/how-to-generate.md)
+
+## Preparing Environment
+In order to generate the Azure CLI, we need 
+1. Nodejs v10+  
+better to be v12+
```

Comment:
could we provide nodejs installation guide with something like below?
`to install NodeJS. You could install a NodeJS globally or use nvm (for linux) or nvm-windows (for windows). It will also help to install NodeJS package management command line tool, npm.
`

Reply:
added